README.md
# scrollable_list_tabview

[![Build Status](https://travis-ci.com/firatcetiner/scrollable_list_tabview.svg?branch=main)](https://travis-ci.com/firatcetiner/scrollable_list_tabview)
[![Codemagic build status](https://api.codemagic.io/apps/5fab163fe55b87e07bbfb1c6/5fab163fe55b87e07bbfb1c5/status_badge.svg)](https://codemagic.io/apps/5fab163fe55b87e07bbfb1c6/5fab163fe55b87e07bbfb1c5/latest_build)
[![Coverage Status](https://coveralls.io/repos/github/firatcetiner/scrollable_list_tabview/badge.svg?branch=main)](https://coveralls.io/github/firatcetiner/scrollable_list_tabview?branch=main)

A Flutter widget which synchronize a ScrollView and a custom tab view.

The main idea is to create a custom tab view synchronizing with inner ScrollView. The scroll activity will trigger custom tab view at the top to follow the index of the inner scroll view widget.


![Demo](https://media2.giphy.com/media/VkrHmPClMEYkBj5Mkg/giphy.gif)

## Installation
Add dependency for package on your pubspec.yaml:

    dependencies:
	    scrollable_list_tabview: <latest>

## Usage
To use this widget we must first define how our tabs will look like.
### ListTab
|Parameter| Definition |
|--|--|
| `Widget icon` |Trailing widget for a tab, typically an Icon.|
|`Widget label`|Label to be shown in the tab, must be non-null.|
| `BorderRadiusGeometry borderRadius`|BorderRadius value for a single tab.|
|`Color activeBackgroundColor`|Color to be used when the tab is selected.|
|`Color inactiveBackgroundColor`| Color to be used when the tab isn't selected.|
|`bool showIconInList`| Decide whether show icon widget in the scrollable view.|
| `Color borderColor`| Color of the border of tab when its not selected|

Then we can use LisTab in ScrollableListTab.
### ScrollableListTab
|Parameter| Definition |
|--|--|
| `ListTab tab` |Data model used for rendering tab widgets.|
|`ScrollView body`|An individual inner scrollable widget.|

Then ScrollableListTabView will take a list of ScrollableListTab as an argument.

### ScrollableListTabView
|Parameter| Definition |
|--|--|
| `List<ScrollableListTab> tabs`|List of tabs to be built.|
|`double height`|Height of the tab at the top of the view.|
| `Duration tabAnimationDuration`|Duration of tab change animation.|
|`Duration bodyAnimationDuration`|Duration of inner scroll view animation.|
| `Curve tabAnimationCurve`|Animation curve used when animating tab change.|
|`Curve bodyAnimationCurve`|Animation curve used when changing index of inner ScrollView(s).|

## Example

    import 'package:flutter/material.dart';
    import 'package:scrollable_list_tabview/scrollable_list_tabview.dart';

    void main() {
      runApp(MyApp());
    }

    class MyApp extends StatelessWidget {
      // This widget is the root of your application.
      @override
      Widget build(BuildContext context) {
        return MaterialApp(
          title: 'Flutter Demo',
          debugShowCheckedModeBanner: false,
          theme: ThemeData(
            primarySwatch: Colors.blue,
            visualDensity: VisualDensity.adaptivePlatformDensity,
          ),
          home: MyHomePage(title: 'Flutter ScrollableListTabView Example'),
        );
      }
    }

    class MyHomePage extends StatefulWidget {
      MyHomePage({Key key, this.title}) : super(key: key);

      final String title;

      @override
      _MyHomePageState createState() => _MyHomePageState();
    }

    class _MyHomePageState extends State<MyHomePage> {
      @override
      Widget build(BuildContext context) {
        return Scaffold(
          appBar: AppBar(
            title: Text(widget.title),
          ),
          body: ScrollableListTabView(
            tabHeight: 48,
            tabs: [
              ScrollableListTab(
                  tab: ListTab(label: Text('Label 1'), icon: Icon(Icons.group)),
                  body: ListView.builder(
                    shrinkWrap: true,
                    physics: NeverScrollableScrollPhysics(),
                    itemCount: 10,
                    itemBuilder: (_, index) => ListTile(
                      leading: Container(
                        height: 40,
                        width: 40,
                        decoration: BoxDecoration(
                            shape: BoxShape.circle, color: Colors.grey),
                        alignment: Alignment.center,
                        child: Text(index.toString()),
                      ),
                      title: Text('List element $index'),
                    ),
                  )),
            ],
          ),
        );
      }
    }




## Limitations & Considerations
There are some limitations on this package. For example we only support ScrollView to be the body. Also, we could have used `builder` pattern to build the widgets.

I would like to thank Google developers for creating awesome package called [ScrollablePositionedList](https://pub.dev/packages/scrollable_positioned_list).


## Contribution
Contributions are accepted via pull requests. For more information about how to contribute to this package, please check the [contribution guide](https://github.com/firatcetiner/scrollable_list_tabview/blob/master/CONTRIBUTION.md).

## License
This project is licensed under the MIT license, additional knowledge about the license can be found [here](https://github.com/firatcetiner/scrollable_list_tabview/blob/master/LICENSE).






CHANGELOG.md
# [0.1.0] - Initial release

### Features
1. Configure background color of labels,
2. Add as many as ScrollView widgets,
3. Synchronize inner scroll view and tab view.


# [0.1.1] - Syntax Changes

This release contains minor syntax changes and new demo gif.

# [0.1.2] - Syntax Changes

Update the documentation.

# [0.1.3] - Minor additions

This release includes the following improvements and features:

1. Add curve parameters for animations of both tab and inner scrollview.
2. Add animation duration parameter for scroll animations of both tab and scrollview.
3. Add necessary error checking assertions.

# [0.1.3+1] - Formatting changes


# [0.1.4] - Syntax improvements

This release includes a new .APK and example shows the usage of curves and animations.

# [0.1.5] - Bug fix

Remove the usage of old OutlineButton and use OutlinedButton instead.

# [0.1.6] - Minor bug fixes

Fix issue with ListTab's parameter of showIconOnList not working properly.


# [0.2.0-nullsafety] - Null Safety

This version opts-in sound null safety for both the package and the example project, along with some minor performance improvements and additional widget tests.



test/scrollable_list_tabview_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:scrollable_list_tabview/scrollable_list_tabview.dart';

void main() {
  testWidgets('Widget tests', (tester) async {
    var listTab = ScrollableListTab(
      tab: ListTab(
        label: const Text('A label'),
        icon: const Icon(Icons.ac_unit),
        showIconOnList: true,
        activeBackgroundColor: Colors.orange,
        borderColor: Colors.grey,
        inactiveBackgroundColor: Colors.transparent,
      ),
      body: ListView(
        shrinkWrap: true,
        physics: const NeverScrollableScrollPhysics(),
      ),
    );
    var widget = ScrollableListTabView(
      tabHeight: 48,
      bodyAnimationDuration: kScrollDuration,
      tabAnimationDuration: kScrollDuration,
      style: const TextStyle(
        color: Colors.grey,
      ),
      tabs: [listTab],
    );
    var testWidget = MaterialApp(
      home: Material(
        child: Directionality(
          textDirection: TextDirection.ltr,
          child: widget,
        ),
      ),
    );
    await tester.pumpWidget(testWidget);
    // Test our main widget.
    expect(find.byWidget(testWidget), findsOneWidget);

    /// Test whether [ScrollableListTabView] has correctly laid out.
    expect(
      find.descendant(
        of: find.byWidget(testWidget),
        matching: find.byType(
          ScrollableListTabView,
        ),
      ),
      findsOneWidget,
    );
    expect(listTab.tab, isA<ListTab>());
    expect(widget.style, isA<TextStyle>());
    expect(widget.style, const TextStyle(color: Colors.grey));
    expect(widget.tabHeight, kTextTabBarHeight);
    expect(widget.tabs.length, 1);

    /// Test animation curves
    expect(widget.tabAnimationCurve, Curves.decelerate);
    expect(widget.bodyAnimationCurve, Curves.decelerate);

    /// Test animation durations
    const duration = const Duration(milliseconds: 150);

    expect(widget.tabAnimationDuration, duration);
    expect(widget.tabAnimationDuration, duration);
  });

  test('Property test for tab models', () {
    var tab = ListTab(label: Text('A label'));
    var scrollableListTab = ScrollableListTab(
      tab: tab,
      body: ListView(
        key: Key('list_view'),
        shrinkWrap: true,
        physics: const NeverScrollableScrollPhysics(),
      ),
    );
    expect(scrollableListTab.tab, tab);
    expect(tab.borderColor, Colors.grey);
    expect(tab.showIconOnList, false);
    expect(tab.icon, null);
    expect(tab.activeBackgroundColor, Colors.blue);
    expect(((tab.label) as Text).data, 'A label');
  });
}


lib/model/list_tab.dart
import 'package:flutter/material.dart';

class ListTab {
  /// Create a new [ListTab]
  const ListTab({
    this.icon,
    required this.label,
    this.borderRadius = const BorderRadius.all(const Radius.circular(5.0)),
    this.activeBackgroundColor = Colors.blue,
    this.inactiveBackgroundColor = Colors.transparent,
    this.showIconOnList = false,
    this.borderColor = Colors.grey,
  });

  /// Trailing widget for a tab, typically an [Icon].
  final Widget? icon;

  /// Label to be shown in the tab, must be non-null.
  final Widget label;

  /// [BorderRadius] for the a tab at the bottom tab view.
  /// This won't affect the tab in the scrollable list.
  final BorderRadiusGeometry borderRadius;

  /// Color to be used when the tab is selected.
  final Color activeBackgroundColor;

  /// Color to be used when tab is not selected
  final Color inactiveBackgroundColor;

  /// If true, the [icon] will also be shown to the user in the scrollable list.
  final bool showIconOnList;

  /// Color of the [Border] property of the inner tab [Container].
  final Color borderColor;
}


lib/scrollable_list_tabview.dart
library scrollable_list_tabview;

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:scrollable_positioned_list/scrollable_positioned_list.dart';
import 'model/scrollable_list_tab.dart';

export 'model/list_tab.dart';
export 'model/scrollable_list_tab.dart';

@visibleForTesting
const Duration kScrollDuration = const Duration(milliseconds: 150);

@visibleForTesting
const EdgeInsetsGeometry kTabMargin =
    const EdgeInsets.symmetric(horizontal: 10.0, vertical: 5.0);

class ScrollableListTabView extends StatefulWidget {
  /// Create a new [ScrollableListTabView]
  const ScrollableListTabView({
    Key? key,
    required this.tabs,
    this.tabHeight = kToolbarHeight,
    this.style = const TextStyle(
      color: Colors.black,
      fontWeight: FontWeight.w500,
    ),
    this.tabAnimationDuration = kScrollDuration,
    this.bodyAnimationDuration = kScrollDuration,
    this.tabAnimationCurve = Curves.decelerate,
    this.bodyAnimationCurve = Curves.decelerate,
  }) : super(key: key);

  /// List of tabs to be rendered.
  final List<ScrollableListTab> tabs;

  /// Height of the tab at the top of the view.
  final double tabHeight;

  /// Duration of tab change animation.
  final TextStyle style;

  /// Duration of tab change animation.
  final Duration tabAnimationDuration;

  /// Duration of inner scroll view animation.
  final Duration bodyAnimationDuration;

  /// Animation curve used when animating tab change.
  final Curve tabAnimationCurve;

  /// Animation curve used when changing index of inner [ScrollView]s.
  final Curve bodyAnimationCurve;

  @override
  _ScrollableListTabViewState createState() => _ScrollableListTabViewState();
}

class _ScrollableListTabViewState extends State<ScrollableListTabView> {
  final ValueNotifier<int> _index = ValueNotifier<int>(0);

  final ItemScrollController _bodyScrollController = ItemScrollController();
  final ItemPositionsListener _bodyPositionsListener =
      ItemPositionsListener.create();
  final ItemScrollController _tabScrollController = ItemScrollController();

  @override
  void initState() {
    super.initState();
    _bodyPositionsListener.itemPositions.addListener(_onInnerViewScrolled);
  }

  @override
  Widget build(BuildContext context) {
    final dividerColor = Theme.of(context).dividerColor;
    return Column(
      children: [
        Container(
          height: widget.tabHeight,
          color: Theme.of(context).cardColor,
          child: ScrollablePositionedList.builder(
            itemCount: widget.tabs.length,
            scrollDirection: Axis.horizontal,
            itemScrollController: _tabScrollController,
            padding: const EdgeInsets.symmetric(vertical: 2.5),
            itemBuilder: (context, index) {
              final tab = widget.tabs[index].tab;
              return ValueListenableBuilder<int>(
                valueListenable: _index,
                builder: (_, i, __) {
                  final selected = index == i;
                  final borderColor =
                      selected ? tab.activeBackgroundColor : dividerColor;
                  return Container(
                    height: 32,
                    margin: kTabMargin,
                    decoration: BoxDecoration(
                      color: selected
                          ? tab.activeBackgroundColor
                          : tab.inactiveBackgroundColor,
                      borderRadius: tab.borderRadius,
                    ),
                    child: OutlinedButton(
                      style: ButtonStyle(
                        foregroundColor: MaterialStateProperty.all<Color>(
                          selected ? Colors.white : Colors.grey,
                        ),
                        backgroundColor: MaterialStateProperty.all<Color>(
                          selected
                              ? tab.activeBackgroundColor
                              : tab.inactiveBackgroundColor,
                        ),
                        tapTargetSize: MaterialTapTargetSize.shrinkWrap,
                        side: MaterialStateProperty.all<BorderSide>(
                          BorderSide(
                            width: 1,
                            color: borderColor,
                          ),
                        ),
                        elevation: MaterialStateProperty.all<double>(0.0),
                        shape:
                            MaterialStateProperty.all<RoundedRectangleBorder>(
                          RoundedRectangleBorder(
                            borderRadius: tab.borderRadius,
                          ),
                        ),
                      ),
                      child: _buildTab(index),
                      onPressed: () => _onTabPressed(index),
                    ),
                  );
                },
              );
            },
          ),
        ),
        Flexible(
          child: ScrollablePositionedList.builder(
            itemScrollController: _bodyScrollController,
            itemPositionsListener: _bodyPositionsListener,
            itemCount: widget.tabs.length,
            itemBuilder: (_, index) => Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              mainAxisSize: MainAxisSize.min,
              children: [
                Padding(
                  padding: kTabMargin.add(const EdgeInsets.all(5.0)),
                  child: _buildInnerTab(index),
                ),
                Flexible(
                  child: widget.tabs[index].body,
                ),
              ],
            ),
          ),
        ),
      ],
    );
  }

  Widget _buildInnerTab(int index) {
    final tab = widget.tabs[index].tab;
    return Builder(
      builder: (_) {
        if (tab.icon == null)
          return tab.label;
        else if (!tab.showIconOnList)
          return DefaultTextStyle(style: widget.style, child: tab.label);
        return DefaultTextStyle(
          style: widget.style,
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            mainAxisSize: MainAxisSize.min,
            children: [_tabIcon(tab.icon), tab.label],
          ),
        );
      },
    );
  }

  Widget _buildTab(int index) {
    final tab = widget.tabs[index].tab;
    return Row(
      crossAxisAlignment: CrossAxisAlignment.center,
      mainAxisAlignment: MainAxisAlignment.center,
      mainAxisSize: MainAxisSize.min,
      children: [_tabIcon(tab.icon), tab.label],
    );
  }

  Widget _tabIcon(Widget? icon) {
    return icon != null
        ? Padding(
            padding: const EdgeInsets.only(right: 8),
            child: icon,
          )
        : const SizedBox(height: 0, width: 0);
  }

  void _onInnerViewScrolled() async {
    final positions = _bodyPositionsListener.itemPositions.value;

    /// Target [ScrollView] is not attached to any views and/or has no listeners.
    if (positions.isEmpty) return;

    /// Capture the index of the first [ItemPosition]. If the saved index is same
    /// with the current one do nothing and return.
    final firstIndex = positions.elementAt(0).index;
    if (_index.value == firstIndex) return;

    /// A new index has been detected.
    await _handleTabScroll(firstIndex);
  }

  Future<void> _handleTabScroll(int index) async {
    _index.value = index;
    await _tabScrollController.scrollTo(
      index: _index.value,
      duration: widget.tabAnimationDuration,
      curve: widget.tabAnimationCurve,
    );
  }

  /// When a new tab has been pressed both [_tabScrollController] and
  /// [_bodyScrollController] should notify their views.
  void _onTabPressed(int index) async {
    _tabScrollController.scrollTo(
      index: index,
      duration: widget.tabAnimationDuration,
      curve: widget.tabAnimationCurve,
    );
    await _bodyScrollController.scrollTo(
      index: index,
      duration: widget.bodyAnimationDuration,
      curve: widget.bodyAnimationCurve,
    );
    _index.value = index;
  }

  @override
  void dispose() {
    _bodyPositionsListener.itemPositions.removeListener(_onInnerViewScrolled);
    return super.dispose();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:scrollable_list_tabview/scrollable_list_tabview.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter ScrollableListTabView Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({
    Key? key,
    required this.title,
  }) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: ScrollableListTabView(
        tabHeight: 48,
        bodyAnimationDuration: const Duration(milliseconds: 150),
        tabAnimationCurve: Curves.easeOut,
        tabAnimationDuration: const Duration(milliseconds: 200),
        tabs: [
          ScrollableListTab(
            tab: ListTab(
              label: Text('Label 1'),
              icon: Icon(Icons.group),
              showIconOnList: false,
            ),
            body: ListView.builder(
              shrinkWrap: true,
              physics: NeverScrollableScrollPhysics(),
              itemCount: 10,
              itemBuilder: (_, index) => ListTile(
                leading: Container(
                  height: 40,
                  width: 40,
                  decoration: BoxDecoration(
                    shape: BoxShape.circle,
                    color: Colors.grey,
                  ),
                  alignment: Alignment.center,
                  child: Text(index.toString()),
                ),
                title: Text('List element $index'),
              ),
            ),
          ),
          ScrollableListTab(
            tab: ListTab(label: Text('Label 2'), icon: Icon(Icons.subject)),
            body: GridView.builder(
              shrinkWrap: true,
              gridDelegate:
                  SliverGridDelegateWithFixedCrossAxisCount(crossAxisCount: 2),
              physics: NeverScrollableScrollPhysics(),
              itemCount: 10,
              itemBuilder: (_, index) => Card(
                child: Center(child: Text('Card element $index')),
              ),
            ),
          ),
          ScrollableListTab(
            tab: ListTab(
                label: Text('Label 3'),
                icon: Icon(Icons.subject),
                showIconOnList: true),
            body: GridView.builder(
              shrinkWrap: true,
              gridDelegate:
                  SliverGridDelegateWithFixedCrossAxisCount(crossAxisCount: 2),
              physics: NeverScrollableScrollPhysics(),
              itemCount: 10,
              itemBuilder: (_, index) => Card(
                child: Center(child: Text('Card element $index')),
              ),
            ),
          ),
          ScrollableListTab(
            tab: ListTab(label: Text('Label 4'), icon: Icon(Icons.add)),
            body: ListView.builder(
              shrinkWrap: true,
              physics: NeverScrollableScrollPhysics(),
              itemCount: 10,
              itemBuilder: (_, index) => ListTile(
                leading: Container(
                  height: 40,
                  width: 40,
                  decoration:
                      BoxDecoration(shape: BoxShape.circle, color: Colors.grey),
                  alignment: Alignment.center,
                  child: Text(index.toString()),
                ),
                title: Text('List element $index'),
              ),
            ),
          ),
          ScrollableListTab(
            tab: ListTab(label: Text('Label 5'), icon: Icon(Icons.group)),
            body: ListView.builder(
              shrinkWrap: true,
              physics: NeverScrollableScrollPhysics(),
              itemCount: 10,
              itemBuilder: (_, index) => ListTile(
                leading: Container(
                  height: 40,
                  width: 40,
                  decoration: BoxDecoration(
                    shape: BoxShape.circle,
                    color: Colors.grey,
                  ),
                  alignment: Alignment.center,
                  child: Text(index.toString()),
                ),
                title: Text('List element $index'),
              ),
            ),
          ),
          ScrollableListTab(
            tab: ListTab(label: Text('Label 6'), icon: Icon(Icons.subject)),
            body: GridView.builder(
              shrinkWrap: true,
              gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
                crossAxisCount: 2,
              ),
              physics: NeverScrollableScrollPhysics(),
              itemCount: 10,
              itemBuilder: (_, index) => Card(
                child: Center(child: Text('Card element $index')),
              ),
            ),
          ),
          ScrollableListTab(
            tab: ListTab(
              label: Text('Label 7'),
              icon: Icon(Icons.subject),
              showIconOnList: true,
            ),
            body: GridView.builder(
              shrinkWrap: true,
              gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                crossAxisCount: 2,
              ),
              physics: NeverScrollableScrollPhysics(),
              itemCount: 10,
              itemBuilder: (_, index) => Card(
                child: Center(child: Text('Card element $index')),
              ),
            ),
          ),
          ScrollableListTab(
            tab: ListTab(label: Text('Label 8'), icon: Icon(Icons.add)),
            body: ListView.builder(
              shrinkWrap: true,
              physics: NeverScrollableScrollPhysics(),
              itemCount: 10,
              itemBuilder: (_, index) => ListTile(
                leading: Container(
                  height: 40,
                  width: 40,
                  decoration:
                      BoxDecoration(shape: BoxShape.circle, color: Colors.grey),
                  alignment: Alignment.center,
                  child: Text(index.toString()),
                ),
                title: Text('List element $index'),
              ),
            ),
          )
        ],
      ),
    );
  }
}
